### PR TITLE
adding powerplatform_connection and powerplatform_connection_share 

### PR DESCRIPTION
**Pull request by mawasile**
This pull request includes significant changes to the Power Platform provider to add support for managing connections and connection shares. The changes include new data sources for fetching connection shares and connections, new resources for creating and managing connections and connection shares, and updates to the provider and testing code to support these new features. The changes are primarily focused on the `powerplatform` package and its sub-packages.

New Data Sources:

* [`docs/data-sources/connection_shares.md`](diffhunk://#diff-8aa36c2d1d6e0e953f7c70ff7e8074561acca60222f55339811054b18bc5b8ecR1-R64): Added a new data source to fetch connection shares. This includes details on how to use the data source and its schema.
* [`docs/data-sources/connections.md`](diffhunk://#diff-4107b66c3e0a7d9e904050fcde33d851014a3ac5143fe006a4d54364358a686cR1-R57): Added a new data source to fetch connections. This includes details on how to use the data source and its schema.

New Resources:

* [`docs/resources/connection.md`](diffhunk://#diff-e00ef64002ad48ae9291de0a2b9fc0147d3820891da4a7ce049c5a705b107600R1-R66): Added a new resource to create and manage connections. This includes details on how to use the resource and its schema.
* [`docs/resources/connection_share.md`](diffhunk://#diff-62667dc4510b4bde7b8d5d47f32ce367fed0dd7fb2cd7386204ef46fb074d957R1-R93): Added a new resource to create and manage connection shares. This includes details on how to use the resource and its schema.

Updates to Provider:

* [`internal/powerplatform/provider.go`](diffhunk://#diff-4ddf04cb6a256a6287ea7f6c6a48a46e29466c3c98d4414dea5bc1f649f6c076R25-R26): Updated the provider to include the new connection and connection share services.

Updates to Testing Code:

* [`internal/powerplatform/datasource_connections_test.go`](diffhunk://#diff-1db8fb7c1fa9997632410aac7bb2bb3b3b5a7262a2dd06d393f9ab6f27b2b122R1-R30): Added a new test for the connections data source.

Examples:

* [`examples/data-sources/powerplatform_connection_shares/`](diffhunk://#diff-986f082469560b16614ceea0345ab64c81bea1787e635237534a538b3485a7c4R1-R18): Added an example of how to use the connection shares data source. [[1]](diffhunk://#diff-986f082469560b16614ceea0345ab64c81bea1787e635237534a538b3485a7c4R1-R18) [[2]](diffhunk://#diff-0a1ca30ddf7d5413ec57265d661a36a7cb305fd8cda30b100bd5a88f6ad81de6R1-R4) [[3]](diffhunk://#diff-33c58834cc59161c271d4755f8bf2aec6cee022bfc62807e50e25e01f5f96a19R1-R12)
* [`examples/data-sources/powerplatform_connections/`](diffhunk://#diff-d6ac6935f591848cfa8fcce043210a958ee77c34e7d89ceb15d248749d249201R1-R17): Added an example of how to use the connections data source. [[1]](diffhunk://#diff-d6ac6935f591848cfa8fcce043210a958ee77c34e7d89ceb15d248749d249201R1-R17) [[2]](diffhunk://#diff-0b86867ab4bea867b1456c96701cc5cef7d7699043df4f466ccafe0ef42f8f0aR1-R4)
* [`examples/resources/powerplatform_connection/`](diffhunk://#diff-3dd1346383442c933c20f12e0c792ef74ecff824220401bd0856eb70f3a9f1cfR1-R31): Added an example of how to use the connection resource. [[1]](diffhunk://#diff-3dd1346383442c933c20f12e0c792ef74ecff824220401bd0856eb70f3a9f1cfR1-R31) [[2]](diffhunk://#diff-a8a5cbe322e5fb8112ce8f3f5726c567929d2a837d5b740ff47504d13bcb6836R1-R6)
* [`examples/resources/powerplatform_connection_share/`](diffhunk://#diff-80da61d7ca8a12455fd60ac10114886a4683fd7d080dd94a351bf5124ac409c0R1-R10): Added an example of how to use the connection share resource. [[1]](diffhunk://#diff-80da61d7ca8a12455fd60ac10114886a4683fd7d080dd94a351bf5124ac409c0R1-R10) [[2]](diffhunk://#diff-7dc58853cb8a71882440fd6a26bad9792288138a4135209da74eeffcd481a0c1R1-R52)

Miscellaneous:

* [`go.mod`](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L38-R38): Updated the `github.com/google/uuid` package to be a direct dependency.
* [`docs/resources/rest.md`](diffhunk://#diff-05ebf31b8a962010445030fefd4f260b65588b56591cb57bd933867e20a7b6eeL6-R14): Fixed some typos in the documentation.

### CODE REVIEW

**Review by eduardodfmex — APPROVED** [review]

File: examples/data-sources/powerplatform_connection_shares/outpuf.tf
Comment (ianjensenisme) [inline examples/data-sources/powerplatform_connection_shares/outpuf.tf:1]: *outpuT.tf

File: examples/data-sources/powerplatform_connection_shares/data-source.tf
Comment (ianjensenisme) [inline examples/data-sources/powerplatform_connection_shares/data-source.tf:1]: Is there a way to get ALL shares across an environment without specifying the connector type, or can I only query the datasource per type?

File: internal/powerplatform/services/connection/api_connection.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+func (client *ConnectionsClient) BuildHostUri(environmentId string) string {
+	envId := strings.ReplaceAll(environmentId, "-", "")
+	realm := string(envId[len(envId)-2:])
+	envId = envId[:len(envId)-2]
+
+	return fmt.Sprintf("%s.%s.environment.%s", envId, realm, client.Api.GetConfig().Urls.PowerPlatformUrl)
+
+}
+
+func (client *ConnectionsClient) GetConnectorDefinition(ctx context.Context, environmentId, connectorName string) (*ConnectorDefinition, error) {
```
Comment (ianjensenisme) [inline internal/powerplatform/services/connection/api_connection.go:38]: This isn't referenced anywhere, what is the intended use case?

File: internal/powerplatform/services/connection/api_connection.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+		return err
+	}
+	return nil
+}
+
+func (client *ConnectionsClient) GetConnectionShares(ctx context.Context, environmentId, connectorName, connectionId string) (*ShareConnectionResponseArrayDto, error) {
+	apiUrl := &url.URL{
+		Scheme: "https",
+		Host:   client.BuildHostUri(environmentId),
+		Path:   fmt.Sprintf("/connectivity/connectors/%s/connections/%s/permissions", connectorName, connectionId),
```
Comment (ianjensenisme) [inline internal/powerplatform/services/connection/api_connection.go:209]: This probably answers my question above..

**Review by ianjensenisme — CHANGES_REQUESTED** [review]
Partial review, I'll try to run it tonight and add more notes. But so far the only "for sure" change request is a typo file rename.

**Comment by ianjensenisme** [comment]
@mawasile I tried to test this locally and in a codespace, and got a mix of different errors in each place, probably related to configuration and Azure security. make test failed with symlink errors, which is probably a container problem. However I think I was able to successfully run the connections datasource example locally, and it looks like it missed the one OAI connection I have in the environment I was testing against. Let's block time to walk through what I'm trying and figure out whether I'm misguided or there's a real issue.

File: internal/powerplatform/datasource_connections_share_test.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+					connection_id  = powerplatform_connection.azure_openai_connection.id
+					role_name      = "CanEdit"
+					principal = {
+					  entra_object_id = azuread_user.test_user.object_id
+					}
+				  }
+				  
+				  data "powerplatform_connection_shares" "all_shares" {
+					environment_id = powerplatform_environment.env.id
+					connector_name = "shared_azureopenai"
```
[the remark this answers is not in the document]
Comment (mawasile) [reply internal/powerplatform/datasource_connections_share_test.go:101]: Why are you assuming that it's not all shares?

File: internal/powerplatform/services/connection/api_connection.go (hunk 1 above)
Comment (mawasile) [reply internal/powerplatform/services/connection/api_connection.go:38]: fair point, removing

**Review by ianjensenisme — APPROVED** [review]

**Review by polatengin — APPROVED** [review]

File: internal/powerplatform/datasource_connections_share_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ianjensenisme) [reply internal/powerplatform/datasource_connections_share_test.go:101]: Line 101 limits the results to just the shared_azureopenai connector, no? But I now realize this is an API limitation, not a problem we need to solve with this PR.

**Pull request merged**